install/infra/terraform/gke/main.tf

```diff
@@ -0,0 +1,191 @@
+terraform {
+  required_version = ">= 1.0.3"
+}
+
+terraform {
+  backend "gcs" {
+    bucket = "gitpod-gke"
+    prefix = "tf-state"
+  }
+}
+
+provider "google" {
+  project     = var.project
+  credentials = var.credentials
+  region      = var.region
+  zone        = var.zone
+}
+
+resource "google_compute_network" "vpc" {
+  name                    = "vpc-${var.name}"
+  auto_create_subnetworks = "false"
+}
+
+resource "google_compute_subnetwork" "subnet" {
+  name          = "subnet-${var.name}"
+  region        = var.region
+  network       = google_compute_network.vpc.name
+  ip_cidr_range = "10.255.0.0/16"
+
+  secondary_ip_range {
+    range_name    = "cluster-secondary-ip-range"
+    ip_cidr_range = "10.0.0.0/12"
+  }
+
+  secondary_ip_range {
+    range_name    = "services-secondary-ip-range"
+    ip_cidr_range = "10.64.0.0/12"
+  }
+}
+
+resource "google_container_cluster" "gitpod-cluster" {
+  name     = "c${var.name}"
+  location = "${var.zone}" == "" ? "${var.region}" : "${var.region}-${var.zone}"
+
+  cluster_autoscaling {
+    enabled = true
+
+    resource_limits {
+      resource_type = "cpu"
+      minimum       = 2
+      maximum       = 8
+    }
+
+    resource_limits {
+      resource_type = "memory"
+      minimum       = 4
+      maximum       = 64
+    }
+  }
+
+  # We can't create a cluster with no node pool defined, but we want to only use
+  # separately managed node pools. So we create the smallest possible default
+  # node pool and immediately delete it.
+  remove_default_node_pool = true
+  initial_node_count       = 1
+  release_channel {
+    channel = "UNSPECIFIED"
+  }
+
+  ip_allocation_policy {
+    cluster_secondary_range_name  = "cluster-secondary-ip-range"
+    services_secondary_range_name = "services-secondary-ip-range"
+  }
+
+  network_policy {
+    enabled  = true
+    provider = "CALICO"
+  }
+
+  addons_config {
+    http_load_balancing {
+      disabled = false
+    }
+
+    horizontal_pod_autoscaling {
+      disabled = false
+    }
+  }
+
+  network    = google_compute_network.vpc.name
+  subnetwork = google_compute_subnetwork.subnet.name
+}
+
+resource "google_container_node_pool" "services" {
+  name               = "services-${var.name}"
+  location           = google_container_cluster.gitpod-cluster.location
+  cluster            = google_container_cluster.gitpod-cluster.name
+  version            = var.kubernetes_version // kubernetes version
+  initial_node_count = 1
+  max_pods_per_node  = 110
```

Comment:
110 is a dangerously high number - if something in the autoscaler goes wrong 110 nodes will lead to high cloud costs. 
Proposal Limit to 6 nodes. That's two per region if this is a regional cluster.

Reply:
Agree that this is quite high, and could be expensive. 6 is pretty reasonable.

For testing purposes I don't think we even need the resilience provided by a regional cluster; if we get a failed CI run due to a GCP zone going offline (which would be an extraordinary event) we can move to a different zone/region. If we opt for a zonal cluster and set the default max to 3 nodes, we can potentially halve our spend for each CI run.

Reply:
> I misread this, this is pods per node. I'm fine with having a high number of pods per node. 

me too :(  110 pods per node is perfectly fine and has zero impact on cost. Apologies for the confusion.